dBildungsplattform/dbildungs-iam-server
# SPSH-89 Keycloak Guards

# Description
This PR adds nest-keycloak-connect which enables verification of users using the authorization header.

To manually test this feature:

1. Create a (or reuse an existing) client in keycloak
2. Make sure the client has "Client authentication" enabled has permission "direct access grants" (getting token on user behalf)
3. Get secret token from "credentials" tab and update server config
4. Request token on behalf of user 
```sh
curl  -X POST \
  'http://127.0.0.1:8080/realms/{{your-realm}}/protocol/openid-connect/token' \
  --header 'Accept: */*' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data-urlencode 'grant_type=password' \
  --data-urlencode 'scope=openid' \
  --data-urlencode 'username={{username}}' \
  --data-urlencode 'password={{password}}' \
  --data-urlencode 'client_id={{client-id}}' \
  --data-urlencode 'client_secret={{client-secret}}'
```
5. Take returned access_token and try protected endpoint
```sh
curl  -X POST \
  'http://127.0.0.1:9091/frontend/logout' \
  --header 'Accept: */*' \
  --header 'Authorization: Bearer {{access_token}}'
```

`/frontend/logout` will only allow authed users to access it



## Links to Tickets or other pull requests
https://ticketsystem.dbildungscloud.de/browse/SPSH-89


## Changes


## Datasecurity


## Deployment


## New Repos, NPM pakages or vendor scripts
- nest-keycloak-connect
  - MIT License


## Approval for review
